### ISSUE TITLE
Some portrait model problems

### ISSUE THREAD

**Issue by jp06** [flowtsohg/mdx-m3-viewer#66]
Now that I have managed to deploy my resources Express server somehow for free, I can open this issue finally. Here's a demo of the website using the model viewer (best having Friz Quadrata TT font installed): https://peaceful-mcnulty-a8a424.netlify.app/ (just search for the respective hero name/title on the search bar on the lower right)

How do I deal with the issue where some models are not visible at all when setting the camera of the portrait model? Most of the models are fine but some models can't be seen at certain aspect ratios of the viewport when it's a bit too wide. Example:
* `Units/Creeps/SylvanusWindrunner/SylvanusWindrunner_portrait.mdx` (search "alleria")
* `Units/Orc/HeroFarSeer/HeroFarSeer_portrait.mdx` (search "chen")
* `Units/NightElf/MountainGiant/MountainGiant_Portrait.mdx` (search "tiny")
* and several others

Also, there's the problem with the camera of some models where the perspective(?) is wrong. Some models I noticed with this problem:
* `Units/Demon/HeroChaosBladeMaster/HeroChaosBladeMaster_portrait.mdx` (search "yurnero")
* `Units/NightElf/DruidoftheClaw/DruidoftheClaw_portrait.mdx` (search "lone druid", only on non-alternate sequences)

I use the following code for setting the perspective and face direction, how the camera values was used was referenced from d07RiV's WC3 Data code:
```javascript
if (model.cameras.length) {
	const { farClippingPlane, fieldOfView, nearClippingPlane, position, targetPosition } = model.cameras[0]
	const aspectRatio = canvasRef.current.width / canvasRef.current.height
	
	// Custom fieldOfView value from d07RiV's WC3 Data model viewer component
	// I don't really understand much but this seems to scale pretty well when used instead of the one in the portrait camera
	const scalableFoV = Math.atan(Math.tan(fieldOfView / 2) / aspectRatio) * 2
	
	instance.resetTransformation()
	instance.setSequence(portraitSequence === -1 ? 0 : portraitSequence)

	scene.camera.perspective(scalableFoV, aspectRatio, nearClippingPlane, farClippingPlane)
	scene.camera.moveToAndFace(position, targetPosition, [0, 0, 1])
}
```

**Comment by flowtsohg** [comment]
I've seen the same issue with some custom models on the Hive recently, still need to debug it properly.

It's also related to the fact the vector spaces in the viewer are messed up, because "forward" means different things to arbitrary nodes than it does to MDX instances than it does to M3 instances.
I am not sure what to do about this which is somewhat blocking me.

**Comment by flowtsohg** [comment]
`setupCamera` now returns a `SimpleOrbitCamera`, which has its own internal `moveToAndFace(position, target)` that updates its internal angles and distance, for smooth transitions.

There was also an issue with instances being culled with the camera being close to them, which made some portraits not render. I made a quick "fix", but I am not sure why it was needed or if it's the right fix, maybe I'll look at this further in the future.

Either way, please test `5.6.8` and see if the issue is solved on your end.

**Comment by jp06** [comment]
I only got to try it just a while ago, and it looks good! A lot of the portrait models that have that culling issue are now visible. You can check it out here: https://wc3-dota-info.netlify.app/.  There are two custom models I noticed that are still not visible (search for "mars" and "apparition").

**Comment by flowtsohg** [comment]
Both have no external portraits, so your portrait code in `loadModels` never runs.

IcyGhost has a `portrait` sequence and the camera to use.

MarsBasic also has a camera, but no portrait sequence. I am assuming the game runs the stand sequences?

**Comment by jp06** [comment]
My code tries to fetch the respective portrait model first. If there's none, it uses the unit model instead and tries to find and set a `portrait` sequence from it. If there's none, it simply just sets it to the first sequence, but now that you mentioned it, setting it to the first stand sequence instead might be better. 

There is this one hero that I don't know what to do with though. The Skeleton King hero doesn't have a portrait model nor a camera on the unit model so I didn't mention it. The map uses `H\SkeletonKing\SkeletonKing.mdx` but there is a similar custom model with a crown on the map in `Units\Undead\Skeleton\Skeleton.mdx` that has a portrait model with it. Is it what's actually being used in some way?

**Comment by flowtsohg** [comment]
So I take it there is still an issue with the mentioned models? I deleted my test code by mistake (github's weird conflict rules strike again), so I can't quickly check it now, will check later.

It would be nice to have screenshots of the in-game portraits for models that don't render for no apparent reason, just so I know what I am supposed to get.

Same for the skeleton king - I played DotA once in my life many years ago, so you'll have to give more data like the model(s?) and portrait screenshot.

**Comment by jp06** [comment]
Yes, the two portrait of the two just simply don't show up. And about Skeleton King, I thought that that it's not necessarily a problem with the viewer and that there's nothing that can be done about it (because no camera/portrait model). But in case you find something: 
[SkeletonKing.zip](https://github.com/flowtsohg/mdx-m3-viewer/files/6133965/SkeletonKing.zip)

And here's the map I extracted the data from: https://www.epicwar.com/maps/308663/

**Comment by jp06** [comment]
Oh forgot the screenshot, it's supposed to look like this: 
![skeleton king](https://user-images.githubusercontent.com/30029476/111015868-c3c2b900-83e5-11eb-81c8-b57d7139164c.PNG)

Which is similar to using `Units\Undead\Skeleton\Skeleton.mdx` instead.

**Comment by jp06** [comment]
Strangely enough, the IcyGhost model portrait views when I check it on the old 6.83d using WC3 Data: https://wc3.rivsoft.net/dota/files/4f519944569dfae2

I don't know the difference except for the fact that they are different files. Also, Skeleton King actually uses the `Units\Undead\Skeleton\Skeleton.mdx` that has a portrait I mentioned earlier.

**Comment by flowtsohg** [comment]
I think I found the real culprit of all of these culling issues - maybe the math wasn't wrong after all.
It happens when a model has no main extent, but does have sequence extents.
As I mentioned in the model centering discussion, the viewer doesn't use the per-sequence extents for culling.
So in the case of not having a main extent, the viewer will do point-culling, and since typically for portraits (or just if you zoom and play with the camera) you won't have the origin visible, the instances are culled.
Need to redesign a bit to use the sequence extents.

I am not sure about the skeleton.
Neither of the models has any kind of portrait as far as I can see (internal, external, imports of which there are none(?), SLKs, etc.).
Additionally, I imported them into `1.29.2`, and this is what it shows:
![Untitled](https://user-images.githubusercontent.com/5874825/111028886-53766080-8402-11eb-8813-e836c7c55f30.jpg)

**Comment by flowtsohg** [comment]
Hopefully that fixes the culling.
Still need to check the IcyGhost face not rendering.

**Comment by flowtsohg** [comment]
And that fixes the face not rendering.

**Comment by jp06** [comment]
Unpublished deploy (model viewer 5.6.10): https://604cdd3e47c75b1687bb249c--wc3-dota-info.netlify.app/
Published deploy (model viewer 5.6.8):  https://wc3-dota-info.netlify.app/

Leaving some feedbacks in the perspective of the above deployed clients:
* Can confirm that the IcyGhost face is now fixed and the Mars model portrait is now visible (IcyGhost portrait still not).
* Updating viewer to 5.6.10, some models are now not working in return like Thrall and Far Seer (search "chen") from 5.6.8.
* Also with model viewer 5.6.10, I noticed that for some reason, it does not render when I switch from another camera view to Portrait view on unit models that have an internal camera and portrait sequences. If you are already in Portrait view though, and you switch to that model the said model from another, it renders fine (unless it doesn't really render from the culling).

**Comment by flowtsohg** [comment]
This entire time I wasn't using the bounds' z coordinate...whoops!
I no longer see IcyGhost and friends randomly disappear when moving the camera in weird angles from close-by.

About switching between views - if this commit doesn't fix it (i.e. it isn't culling issues), can you try to use the camera code from [camera.js](https://github.com/flowtsohg/mdx-m3-viewer/blob/master/clients/shared/camera.js)? I am not quite sure how your current camera code looks.

**Comment by jp06** [comment]
5.6.12 - https://wc3-dota-info.netlify.app/
5.6.10 - https://604cdd3e47c75b1687bb249c--wc3-dota-info.netlify.app/
5.6.8 - https://604e14202736280007f798c3--wc3-dota-info.netlify.app/

Looks great! I think the switching views strange behavior only happens right now on the latest model viewer version with Zet only (Gnoll Warden model), strangely enough. Back on 5.6.10, I noticed that it also happens with the Lucifer (Doom Bringer), Mars, and Riki (Satyr Trickster) models and maybe others that I haven't mentioned with an internal camera. 

I simply just added an option on the camera to lock vertical rotation for use on the Side View. I also tried your new camera.js code and also added the vertical rotation lock option to it and it seems to be behaving the same.

**Comment by flowtsohg** [comment]
Since the portraits now show when they should when in portrait mode, the issue seems to be something the client is doing.
I looked a bit at your code, but couldn't tell what's the difference between selecting a model with the mode being in portrait, and selecting the portrait mode. Never used React or its JSX format, so I am having trouble following the code 🤔

**Comment by jp06** [comment]
I'm not sure if it's a problem with the client as far as I'm aware, I simply just updated the viewer from 5.6.8 when it started happening. I believe it would go wrong from 5.6.8 and below already if it is a problem with my current code. It just runs this part every time a new camera view is selected (didn't copy the whole thing):

```javascript
scene.camera.reset()

switch (newCameraView) {
	case 0: // Close-up
	case 2: // Free	
		const xAxisRotationOnlyOptions = {
			horizontalAngle: Math.PI / 1.5,
			verticalAngle: Math.PI / 2,
			distance: 330,
			lockVerticalRotate: true
		}

		unitInstance.setScale(scale)
		unitInstance.setUniformScale(modelScale)
		unitInstance.setLocation([0, 0, -100 + (moveHeight ? moveHeight / 2 : 0)]);
		unitInstance.setSequence(firstStandSequence);
		unitInstance.show()

		if (portraitInstance) portraitInstance.hide()
		
		setupCamera(scene, newCameraView === 0 ? xAxisRotationOnlyOptions : {})
		setSequence(firstStandSequence)
		
		break
	case 1: // Portrait
		const instance = portraitInstance ? portraitInstance : unitInstance
		const model = portraitInstance ? portraitModel : unitModel
		const portraitSequence = portraitInstance && newCameraView === 1
			? model.name === "HeroMountainKing_Portrait" 
				? 5
				: model.sequences.findIndex(({name}) => name.toLowerCase().includes("portrait"))
			: model.sequences.findIndex(({name}) => name.toLowerCase().includes("portrait"))

		if (model.cameras.length) {
			const { farClippingPlane, fieldOfView, nearClippingPlane, position, targetPosition } = model.cameras[0]
			const aspectRatio = canvasRef.current.width / canvasRef.current.height
			const scalableFoV = Math.atan(Math.tan(fieldOfView / 2) / aspectRatio) * 2
			
			instance.resetTransformation()
			instance.setSequence(portraitSequence === -1 ? 0 : portraitSequence)

			scene.camera.perspective(scalableFoV, aspectRatio, nearClippingPlane, farClippingPlane)
			scene.camera.moveToAndFace(position, targetPosition, [0, 0, 1])
		}
		
		if (portraitInstance) {
			unitInstance.hide()
			portraitInstance.show()
		} else {
			unitInstance.show()
		}
		
		setSequence(portraitSequence === -1 ? 0 : portraitSequence)

		break
	default:
		console.error("Invalid camera view index of " + newCameraView)
}
```

I'm not sure what has changed from 5.6.10 to 5.6.12 but it's definitely more stable in terms of that certain issue. I haven't tried with 5.6.9 though because I jumped to 5.6.10, what if that's actually where it all started.

**Comment by jp06** [comment]
I think it views after a `scene.clear()` before setting the portrait view (which is a 100% success). It runs once on my code when I load new models. Which should be why it loads when switching from another model to the one that doesn't load when changing from other views to Portrait view.

And is something I think is worth mentioning, it also views after an `instance.setSequence()` to certain sequences which I found out recently. In the Gnoll Warden model, while in Portrait view, when I change to either `Spell`, `Stand - 4`, `Attack - 1` or `Attack - 2` sequences, it would successfully view and it would stay the same after you return to other stand or portrait sequences. 

Probably something about sequences that causes it? I was thinking that maybe `scene.camera.reset()` doesn't reset some stuff.

**Comment by flowtsohg** [comment]
The only issue I could see here is calling `setupCamera` multiple times.
The class it returns registers event listeners in the constructor, and you're not really meant to run it on the same scene again.

I honestly didn't remember `Camera.reset()` existed, but it doesn't really matter here - in any case `Camera.moveToAndFace()` will be called, either by `camera.js` or by you.

Is there some way for me to run this code locally so I can test things?

**Comment by jp06** [comment]
I just recalled that it's something I actually have to address. I have to set the camera in my HeroModelViewer component instead of using external stuff but `camera.js` seems to do the job pretty well even with the redundant adding of listeners so I've been putting it off. 

---

### Running locally

Resource Server:
1. Clone `https://github.com/jp06/wc3-dota-info-api.git`
2. Download this [sample assets](https://github.com/jp06/wc3-dota-info-api/releases/download/v1.0/resources.7z) and extract and merge to `resources/` folder in the directory
3. `npm install`
4. `npm start`

Client:
1. Clone `https://github.com/jp06/wc3-dota-info.git`
2. `npm install`
3. `npm start`
4. Open http://localhost:8080

PS: I haven't actually tried it.

**Comment by flowtsohg** [comment]
The Gnoll Warden is at `Z = -100` when switching from free mode to portrait.
You then call `resetTransformation()` which sets `Z = 0`, however it only sets the local location.
The world location is updated later when the instance gets updated.
Then the viewer checks if the instance is visible, which uses the world location which still has `Z = -100`, determines the instance isn't visible, and doesn't update it.
Any kind of forced update for the Z value (`update()`, `recalculateTransformation()`, or even setting `instance.worldLocation[2] = 0`) will solve this.

I was actually just today thinking about whether it's right for the camera class to have a dirty state and whether it should instead update instantly when anything changes.
It means that if you want to do multiple changes per frame it's extra work, however if the values are needed this frame, it's somewhat problematic.

The same can be said for nodes.
Skeletal nodes do need their own dirty state and to be optimized more, but for normal nodes, which instances are, maybe it makes more sense to update instantly.
So e.g. when you run `instance.resetTransformation()`, `instance.worldLocation[2]` will actually be 0.

**Comment by jp06** [comment]
Oh, that's the case, I see. So I read about dirty checking today, and now I can kind of understand what that `dirty` boolean value is for.

For now, I just added `recalculateTransformation()` and that seems to fix it.
Thanks for the explanation, really appreciate it.

Out of curiosity, what's the difference between `worldLocation` and `location`? I assume `worldLocation` is something like global and the `location` is local but other than that, what's the difference?

**Comment by flowtsohg** [comment]
You shouldn't need to call `recalculateTransformation()`, your code should intuitively work, so as far as I am concerned this is bug, but if you want a fix for now that will work.

`localX` is in object space, and `worldX` is in world space.
If a node isn't parented they are the same, but if a node is parented, the local values are relative to the parent, while the world ones are absolute.

**Comment by flowtsohg** [comment]
I chose to not change how nodes work, because I will need to think about the implications of that a lot more, since it affects the entire viewer, and it might introduce performance issues.
Instead instances that are dirty (i.e. the node changed) will now update even if they are not visible, meaning the client can move instances in any way it likes and everything should work...at least as long as the instances don't go out of the grid, but that typically shouldn't happen.

I might still change the camera so it's updated instantly and isn't a part of `Scene.update()`. It makes sense, since the camera is meant to be controlled by the client.

**Comment by jp06** [comment]
Sweet! Now what's left are the ones that doesn't view incorrectly.

These are the models I noticed with this behavior so far that I remember:
* `Units\Demon\HeroChaosBladeMaster\HeroChaosBladeMaster_Portrait.mdx`
* `Units\NightElf\DruidoftheClaw\DruidoftheClaw_Portrait.mdx` (alternate portrait sequences)
* `Units\Orc\HeroTaurenChieftain\HeroTaurenChieftain_Portrait.mdx`
* `Units\Orc\SpiritWalker\SpiritWalker_Portrait.mdx`

**Comment by flowtsohg** [comment]
I checked the Tauren Chieftain, and the camera has its own animations that need to be taken into consideration.

You'll need something like the following code, however note that the MDX camera class is a bit bugged and I need to commit a fix.

```ts
let camera = model.cameras[0];

let position = camera.position;
let translation = [];

camera.getTranslation(translation, sequence, frame, counter); // the last 3 parameters can be taken directly from the instance

let finalPosition = [
  position[0] + translation[0],
  position[1] + translation[1],
  position[2] + translation[2]
];
```

Same goes with the target translation.

It also has a rotation animation, but I actually never looked at those until now.
To my knowledge, a camea's rotation animation is a single angle that affects the camera's twist, however I just looked and the data type looks wrong.
I'll need to look more at this.

**Comment by flowtsohg** [comment]
I've incorporated this new code into SimpleOrbitCamera as a test.

First in general you can now control the twist by changing `camera.twist` and then updating the camera.

Second you can call `camera.applyInstanceCamera(instance, cameraIndex)` to change to a mode where the orbit camera isn't actually an orbit camera and instead follows the given MDX camera (via index which will usually be 0).
If the client changes the camera manually with the mouse or touch events, then the camera will stop following the MDX camera, and if you set `SimpleOrbitCamera.onManualChange` to a function (manually or via `options` in the constructor), it will be called so you can update UI or whatever you want.

Please test it :P

**Comment by jp06** [comment]
Hi, I just tested it again. Now the only one that still doesn't view correctly is the Druid of the Claw - Night Elf Form sequences. I've been saying alternate sequences all this time but it was actually the Bear Form sequences that are the alternate 😅. I tweaked mine a bit to use a horizontal FoV ([formula](https://en.wikipedia.org/wiki/Field_of_view_in_video_games#Field_of_view_calculations)).

I've been trying to implement an inhouse camera like how d07RiV has implemented [here](https://github.com/d07RiV/wc3data/blob/master/src/files/FileModel.js). Still no success, I'm not sure how to port his camera code on mine. I never checked a problematic portrait on his viewer until I got curious yesterday and it seems that he already solved the portrait problem and even the [Druid of the Claw](https://wc3.rivsoft.net/7085/files/8c9bfc79836eeb65) renders correctly.

**Comment by flowtsohg** [comment]
It looks fine for me as far as I can tell using `SimpleOrbitCamera.applyInstanceCamera`:
![Untitled](https://user-images.githubusercontent.com/5874825/113081931-4b7c3600-91e2-11eb-9d91-659bee2eb6b3.jpg)

I'll be honest, I am not quite sure what's the purpose of the dynamic FOV, but when using a model camera you should use its own FOV regardless.

**Comment by jp06** [comment]
I think for now I'll try tweaking my code a bit more to make it work.

I actually first tried using the actual FoV of the portrait camera back then. Only, the aspect ratio of my canvas is a bit wide so I end up seeing a lot of decapitated arm/body/head on a lot of models when applying that FoV. I think it would be fine if the width is not greater than the height of the canvas but that's not the case in my project. Generally, I find the calculated horizontal FoV better for my certain project, the decapitated arms are now hidden/less noticeable.

I used this code on the `applyInstanceCamera()` method:
```javascript
const aspectRatio = this.canvas.width / this.canvas.height
const camFoV = instance.model.cameras[cameraIndex].fieldOfView
const horizontalFoV = Math.atan(Math.tan(camFoV / 2) / aspectRatio) * 2

this.fov = horizontalFoV
```
Here are a few comparisons of the two FoV:
![ember_cam_fov](https://user-images.githubusercontent.com/30029476/113093644-e6950000-9222-11eb-9d6d-b18567b73e08.PNG) 
![ember_alt_fov](https://user-images.githubusercontent.com/30029476/113093637-e4cb3c80-9222-11eb-97d5-f6958351fcb0.PNG)

![lion_alt_fov](https://user-images.githubusercontent.com/30029476/113093648-e72d9680-9222-11eb-939f-032a8beab5dc.PNG)
![lion_cam_fov](https://user-images.githubusercontent.com/30029476/113093652-e7c62d00-9222-11eb-816e-b146f248b664.PNG)

![phoenix_alt_fov](https://user-images.githubusercontent.com/30029476/113093654-e85ec380-9222-11eb-81fa-6134aec49620.PNG)
![phoenix_cam_fov](https://user-images.githubusercontent.com/30029476/113093656-e8f75a00-9222-11eb-90d1-e3c870cc1bb9.PNG)

![techies_alt_fov](https://user-images.githubusercontent.com/30029476/113093659-e98ff080-9222-11eb-98ba-c8fab3a8221c.PNG)
![techies_cam_fov](https://user-images.githubusercontent.com/30029476/113093662-eac11d80-9222-11eb-98b0-6c1941392582.PNG)

![wr_alt_fov](https://user-images.githubusercontent.com/30029476/113093666-ebf24a80-9222-11eb-90cb-178fb441bcb6.PNG)
![wr_cam_fov](https://user-images.githubusercontent.com/30029476/113093669-ed237780-9222-11eb-88c1-ec32ffc4cbb8.PNG)

![zeus_alt_fov](https://user-images.githubusercontent.com/30029476/113093670-edbc0e00-9222-11eb-8270-fdcbcdab5e02.PNG)
![zeus_cam_fov](https://user-images.githubusercontent.com/30029476/113093672-eeed3b00-9222-11eb-8300-69417f2b968f.PNG)

**Comment by flowtsohg** [comment]
I don't think the different FOV is resulting in more correct rendering. You are cutting out parts that aren't visible in the game, but you are doing so by zooming in rather than having an overlay, it's not quite the same operation.

As a side note, if you want for example to use the canvas as a square canvas, even if it isn't, you can use `Scene.viewport` for a letterboxin effect.
The one caveat is `SimpleOrbitCamera.onResize()` sets the viewport automatically to the entire canvas, that probably needs to change.
For example, suppose your canvas is 300x100, and you want to render on 100x100 with black strips on the sides:
```ts
let viewport = scene.viewport;

viewport[0] = 100; // offset from the left
viewport[2] = 100; // width
```

**Comment by jp06** [comment]
I suppose using whichever one is subjective. I'm not saying what I used is more correct or proper though. Like what I said, I just prefer it. I think the difference is kinda like this with a wide aspect ratio?

![fov-comparison](https://user-images.githubusercontent.com/30029476/113252292-38f32100-92f6-11eb-9cc4-880c2c19d0b4.jpg)

I appreciate the suggestion about the letterboxing, I've been thinking about something similar to try replicate like what's only shown on WC3 portrait but I'll still opt to use the zoomed fov with the cut top and bottom for now. I just feel like it won't look as good if I did that, maybe in the future I'll try to think of something to have a better portrait view.
